File: src/main/java/com/ballgame/Circle.java
package com.ballgame;

public abstract class Circle extends GameObject {
    protected int r;

    public Circle(int x, int y) {
        super(x, y);
    }

    public double getCenterX() {
        return this.x + this.r;
    }

    public int getRadius() {
        return this.r;
    }

    public int getDiameter() {
        return this.r * 2;
    }

    public double getCenterY() {
        return this.y + this.r;
    }

    public boolean intersects(Circle other) {
        double distance = this.getDistance(this.getCenterX(), this.getCenterY(), other.getCenterX(),
                other.getCenterY());
        ;
        return distance <= this.r + other.r;
    }
}

File: src/main/kotlin/com/ballgame/Circle.kt
package com.ballgame

abstract class Circle(x: Int, y: Int) : GameObject(x, y) {
	var radius = 0
		protected set
	private val centerX: Double
		get() = this.x + radius
	val diameter: Int
		get() = radius * 2
	private val centerY: Double
		get() = this.y + radius

	fun intersects(other: Circle) = this.getDistance(
		centerX, centerY, other.centerX, other.centerY
	) <= radius + other.radius
}
